cpp/src/text/edit_distance.cu

```diff
@@ -0,0 +1,316 @@
+  // we can use the output column buffer to hold the size/offset values temporarily
+  auto results   = cudf::make_fixed_width_column(cudf::data_type{cudf::type_id::INT32},
+                                               strings_count,
+                                               rmm::device_buffer{0, stream, mr},
+                                               0,
+                                               stream,
+                                               mr);
+  auto d_results = results->mutable_view().data<int32_t>();
+  auto execpol   = rmm::exec_policy(stream);
+  thrust::transform(execpol->on(stream),
+                    thrust::make_counting_iterator<cudf::size_type>(0),
+                    thrust::make_counting_iterator<cudf::size_type>(strings_count),
+                    d_results,
+                    [d_strings, d_targets] __device__(auto idx) {
+                      if (d_strings.is_null(idx) || d_targets.is_null(idx)) return int32_t{0};
+                      auto d_str = d_strings.element<cudf::string_view>(idx);
+                      auto d_tgt = d_targets.size() == 1
+                                     ? d_targets.element<cudf::string_view>(0)
+                                     : d_targets.element<cudf::string_view>(idx);
+                      // just need 3 int16's for each character of the shorter string
+                      return static_cast<int32_t>(std::min(d_str.length(), d_tgt.length()) * 3);
+                    });
+
+  // get the total size of the temporary compute buffer
+  size_t compute_size =
+    thrust::reduce(execpol->on(stream), d_results, d_results + strings_count, size_t{0});
+  // convert sizes to offsets in-place
+  thrust::exclusive_scan(execpol->on(stream), d_results, d_results + strings_count, d_results);
+  // create the temporary compute buffer
+  rmm::device_uvector<int16_t> compute_buffer(compute_size, stream);
+  auto d_buffer = compute_buffer.data();
+
+  // compute the edit distance into the output column in-place
+  // - on input, d_results is the offset to the working section of d_buffer for each row
+  // - on output, d_results is the calculated edit distance for that row
+  thrust::for_each_n(
+    execpol->on(stream),
+    thrust::make_counting_iterator<cudf::size_type>(0),
+    strings_count,
+    edit_distance_levenshtein_algorithm{d_strings, d_targets, d_buffer, d_results});
+  return results;
+}
+
+/**
+ * @copydoc nvtext::edit_distance_matrix
+ */
+std::unique_ptr<cudf::column> edit_distance_matrix(cudf::strings_column_view const& strings,
+                                                   cudaStream_t stream,
+                                                   rmm::mr::device_memory_resource* mr)
+{
+  cudf::size_type strings_count = strings.size();
+  if (strings_count == 0) return cudf::make_empty_column(cudf::data_type{cudf::type_id::INT32});
+  CUDF_EXPECTS(strings_count > 1, "the input strings must include at least 2 strings");
```

Comment:
No scalar?

Reply:
The Python/Cython code already had a logic/utility to turn an input string into a column so I did not see a need to create a scalar-specific version of the API since Python would not need it. Also, libcudf has a `make_column_from_scalar` utility if a C++ user wishes to use it.